File: src/main/java/waterpower/common/block/watermills/TileEntityWatermill.java
package waterpower.common.block.watermills;

import net.minecraft.block.state.IBlockState;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import waterpower.WaterPower;
import waterpower.client.gui.DefaultGuiIds;
import waterpower.common.block.tileentity.TileEntityRotor;
import waterpower.common.item.range.ItemRange;
import waterpower.common.item.range.RangeInventorySlot;
import waterpower.common.item.range.RangeType;
import waterpower.util.Utils;

/**
 * 
 * @author jackhuang1998
 * 
 */
public class TileEntityWatermill extends TileEntityRotor {

    private boolean canWheelTurn = false;
    public int waterBlocks, lavaBlocks;
    public int preWaterBlocks = -999, preLavaBlocks = -999;
    public int range = 0, rotor = -1;

    RangeInventorySlot slotUpdater;
    
    {
        slotUpdater = new RangeInventorySlot(this);
        addInvSlot(slotUpdater);
    }

    public TileEntityWatermill() {
        super();
    }
    
    public TileEntityWatermill(WaterType type) {
    	super(type.output, 2097152);
    	temporaryType = type;
    }
    
    @Override
    protected void func_190201_b(World p_190201_1_) {
    	super.func_190201_b(p_190201_1_);
    	worldObj = p_190201_1_;
    	getType();
    }
    
    @Override
    public IBlockState getBlockState() {
    	return super.getBlockState().withProperty(BlockWatermill.WATERMILL_TYPES, getType());
    }

    @Override
    public void writePacketData(NBTTagCompound tag) {
        super.writePacketData(tag);
        tag.setInteger("waterBlocks", waterBlocks);
        tag.setInteger("lavaBlocks", lavaBlocks);

        rotor = hasRotor() ? getRotor().type.ordinal() : -1;
        tag.setInteger("rotor", rotor);
        tag.setInteger("range", range);
    }

    @Override
    public void readPacketData(NBTTagCompound tag) {
        super.readPacketData(tag);
        waterBlocks = tag.getInteger("waterBlocks");
        lavaBlocks = tag.getInteger("lavaBlocks");

        rotor = tag.getInteger("rotor");
        range = tag.getInteger("range");
    }
    
    WaterType temporaryType = null;
    
    public WaterType getType() {
    	if (temporaryType != null) return temporaryType;
    	temporaryType = WaterType.values()[getBlockMetadata()];
        production = temporaryType.output;
        maxStorage = 2097152;
        return temporaryType;
    }

    private void getWaterBlocks() {
        preLavaBlocks = lavaBlocks;
        preWaterBlocks = waterBlocks;
        lavaBlocks = waterBlocks = 0;

        int range = getRange();
        if (range * range * range > 729) {
            lavaBlocks = waterBlocks = -1;
            return;
        }
        range /= 2;

        for (int xTest = -range; xTest <= range; xTest++) {
            for (int yTest = -range; yTest <= range; yTest++) {
                for (int zTest = -range; zTest <= range; zTest++) {
                	BlockPos newPos = pos.add(xTest, yTest, zTest);
                    if (Utils.isWater(worldObj, newPos)) {
                        waterBlocks++;
                    } else if (Utils.isLava(worldObj, newPos)) {
                        lavaBlocks++;
                    }
                }
            }
        }

        if (getType().ordinal() < 2)
            lavaBlocks = 0;
    }

    public boolean isRangeSupported() {
        return waterBlocks != -1 && lavaBlocks != -1;
    }

    @Override
    protected double computeOutput(World world, BlockPos pos) {
        if (waterBlocks == -1) {
            return 0;
        }

        double vol = getRange();
        vol *= vol * vol;
        double waterPercent = waterBlocks / (vol - 1);
        double lavaPercent = lavaBlocks / (vol - 1);
        double percent = waterPercent + lavaPercent * 4;
        double energy = getType().output * percent;
        energy *= tickRotor();
        if (energy > 0)
            damageRotor(1);
        return energy;
    }

    @Override
    protected int getProduction() {
        if (lavaBlocks > 0)
            return super.getProduction() * 4;
        else
            return super.getProduction();
    }

    public boolean hasRangeUpdater() {
        if (slotUpdater != null && !slotUpdater.isEmpty()) {
            for (int i = 0; i < slotUpdater.size(); i++)
                if (slotUpdater.get(i) != null && slotUpdater.get(i).getItem() instanceof ItemRange)
                    return true;
        }
        return false;
    }

    public int getRange() {
        if (WaterPower.isClientSide() && range != 0)
            return range;
        int range = getType().length;
        if (slotUpdater != null && !slotUpdater.isEmpty()) {
            for (int i = 0; i < slotUpdater.size(); i++) {
                ItemStack is = slotUpdater.get(i);
                if (is == null)
                    continue;
                if (is != null && slotUpdater.get(i).getItem() instanceof ItemRange && is.getItemDamage() >= RangeType.values().length)
                    return getType().length;
                range -= is.stackSize * RangeType.values()[is.getItemDamage()].range;
            }
        }
        if (range < 3)
            range = 3;
        return range;
    }

    @Override
    public String getName() {
        return getType().getShowedName();
    }

    @Override
    public int getGuiId() {
        return DefaultGuiIds.get("tileEntityWatermill");
    }

    @Override
    protected void onUpdate() {
        super.onUpdate();

        getWaterBlocks();
    }

    @Override
    public boolean isActive() {
        return super.isActive() && (storage < maxStorage);
    }

    @Override
    protected boolean allowedSendPacketTank() {
        return false;
    }
}

File: src/main/kotlin/waterpower/common/block/watermill/TileEntityWatermill.kt
/**
 * Copyright (c) Huang Yuhui, 2017
 *
 * "WaterPower" is distributed under the terms of the Minecraft Mod Public
 * License 1.0, or MMPL. Please check the contents of the license located in
 * http://www.mod-buildcraft.com/MMPL-1.0.txt
 */
package waterpower.common.block.watermill

import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import waterpower.annotations.HasGui
import waterpower.annotations.Register
import waterpower.annotations.Sync
import waterpower.common.block.inventory.InventorySlotRange
import waterpower.common.block.tile.TileEntityRotorGenerator
import waterpower.common.init.WPBlocks
import waterpower.common.item.ItemRange
import waterpower.common.item.RangePlugins
import waterpower.util.isLava
import waterpower.util.isWater

@HasGui(guiClass = GuiWatermill::class, containerClass = ContainerWatermill::class)
open class TileEntityWatermill(val type: EnumWatermill) : TileEntityRotorGenerator(type.getOutput(), 2097152.0, type.getTier()) {
    val slotUpdater = InventorySlotRange(this)

    @Sync var waterBlocks: Int = 0
    @Sync var lavaBlocks: Int = 0
    @Sync var _range = 0

    override fun getName() = type.getLocalizedName()

    override fun getBlockState() = super.getBlockState().withProperty(WPBlocks.watermill.TYPES, type)!!

    private fun getWaterBlocks() {
        waterBlocks = 0
        lavaBlocks = waterBlocks

        var range = getRange()
        if (range * range * range > 729) {
            waterBlocks = -1
            lavaBlocks = waterBlocks
            return
        }
        range /= 2

        for (xTest in -range..range) {
            for (yTest in -range..range) {
                for (zTest in -range..range) {
                    val newPos = pos.add(xTest, yTest, zTest)
                    if (isWater(getWorld(), newPos)) {
                        waterBlocks++
                    } else if (isLava(getWorld(), newPos)) {
                        lavaBlocks++
                    }
                }
            }
        }

        if (type.ordinal < 2)
            lavaBlocks = 0
    }

    override fun computeOutput(world: World, pos: BlockPos): Double {
        if (waterBlocks == -1) {
            return 0.0
        }

        var vol = getRange()
        vol *= vol * vol
        val waterPercent = waterBlocks * 1.0 / (vol - 1)
        val lavaPercent = lavaBlocks * 1.0 / (vol - 1)
        val percent = waterPercent + lavaPercent * 4
        var energy = type.getOutput() * percent
        energy *= getEfficiency()
        if (energy > 0)
            damageRotor(1)
        return energy
    }

    override fun onUpdate() {
        super.onUpdate()

        getWaterBlocks()
    }

    fun isRangeSupported() = waterBlocks != -1 && lavaBlocks != -1

    fun hasRangeUpdater(): Boolean {
        if (!slotUpdater.isEmpty()) {
            for (i in 0..slotUpdater.size() - 1)
                if (!slotUpdater.get(i).isEmpty && slotUpdater.get(i).getItem() is ItemRange)
                    return true
        }
        return false
    }

    fun getRange(): Int {
        if (getWorld().isRemote && _range != 0)
            return _range
        _range = type.checkSize
        if (hasRangeUpdater()) {
            for (i in 0..slotUpdater.size() - 1) {
                val stack = slotUpdater.get(i)
                if (stack.isEmpty) continue
                if (slotUpdater.get(i).getItem() is ItemRange && stack.getItemDamage() >= RangePlugins.values().size)
                    return type.checkSize
                _range -= stack.count * RangePlugins.values()[stack.getItemDamage()].range
            }
        }
        if (_range < 3)
            _range = 3
        return _range
    }
}

@Register("waterpower.watermill.mk1")
class TileEntityWatermillMK1 : TileEntityWatermill(EnumWatermill.MK1)

@Register("waterpower.watermill.mk2")
class TileEntityWatermillMK2 : TileEntityWatermill(EnumWatermill.MK2)

@Register("waterpower.watermill.mk3")
class TileEntityWatermillMK3 : TileEntityWatermill(EnumWatermill.MK3)

@Register("waterpower.watermill.mk4")
class TileEntityWatermillMK4 : TileEntityWatermill(EnumWatermill.MK4)

@Register("waterpower.watermill.mk5")
class TileEntityWatermillMK5 : TileEntityWatermill(EnumWatermill.MK5)

@Register("waterpower.watermill.mk6")
class TileEntityWatermillMK6 : TileEntityWatermill(EnumWatermill.MK6)

@Register("waterpower.watermill.mk7")
class TileEntityWatermillMK7 : TileEntityWatermill(EnumWatermill.MK7)

@Register("waterpower.watermill.mk8")
class TileEntityWatermillMK8 : TileEntityWatermill(EnumWatermill.MK8)

@Register("waterpower.watermill.mk9")
class TileEntityWatermillMK9 : TileEntityWatermill(EnumWatermill.MK9)

@Register("waterpower.watermill.mk10")
class TileEntityWatermillMK10 : TileEntityWatermill(EnumWatermill.MK10)

@Register("waterpower.watermill.mk11")
class TileEntityWatermillMK11 : TileEntityWatermill(EnumWatermill.MK11)

@Register("waterpower.watermill.mk12")
class TileEntityWatermillMK12 : TileEntityWatermill(EnumWatermill.MK12)

@Register("waterpower.watermill.mk13")
class TileEntityWatermillMK13 : TileEntityWatermill(EnumWatermill.MK13)

@Register("waterpower.watermill.mk14")
class TileEntityWatermillMK14 : TileEntityWatermill(EnumWatermill.MK14)

@Register("waterpower.watermill.mk15")
class TileEntityWatermillMK15 : TileEntityWatermill(EnumWatermill.MK15)

@Register("waterpower.watermill.mk16")
class TileEntityWatermillMK16 : TileEntityWatermill(EnumWatermill.MK16)